src/groth16/circuit/builder/mod.rs

```diff
@@ -72,13 +90,266 @@ where
+    /// circuit.set_word8(&u8_input, 0b0000_0010);
+    /// assert_eq!(circuit.evaluate_word8(&u8_input), 0b0000_0010);
+    ///
+    /// // ...
+    /// ```
+    pub fn new_word8(&mut self) -> Word8 {
+        let mut wrd8: Word8 = Word8::default();
+        (0..8).for_each(|x| wrd8[x] = self.new_wire());
+        wrd8
+    }
+
+    pub fn new_word8_array<'a>(&mut self, output: &'a mut [Word8]) {
+        (0..output.len()).for_each(|i| output[i] = self.new_word8());
+    }
+
+    pub fn new_word8_vec(&mut self, size: usize) -> Vec<Word8> {
+        (0..size).map(|_| self.new_word8()).collect()
+    }
+
+    /// Creates a new u64 "number", but this is not the right way to think about
+    /// it. Really it a conduit that accepts a u64 number as input where the
+    /// wire numbers correspond to the bits of the u64 number. You can almost
+    /// think of it as a type for circuits since it constrains the input of
+    /// circuit builders. However, circuits at this level have no type, so there
+    /// is nothing to prevent you from inputing numbers other than 0 or 1 in the
+    /// wire inputs.
+    ///
+    /// ```
+    /// use zksnark::field::z251::Z251;
+    /// use zksnark::field::*;
+    /// use zksnark::groth16::circuit::*;
+    ///
+    /// // Create an empty circuit
+    /// let mut circuit = Circuit::<Z251>::new();
+    ///
+    /// let u64_input = circuit.new_word64();
+    ///
+    /// circuit.set_word64(&u64_input, 1);
+    ///
+    /// assert_eq!(circuit.evaluate_word64(&u64_input), 1);
+    /// ```
+    pub fn new_word64(&mut self) -> Word64 {
+        let mut wrd64: Word64 = Word64::default();
+        (0..8).for_each(|x| wrd64[x] = self.new_word8());
+        wrd64
+    }
+
+    /// NOTE: this is only used internally to implement Keccak
+    fn initial_keccakmatrix(&mut self) -> [Word64; 25] {
+        [Word64::default(); 25]
+    }
+
+    // NOTE: this is only used for testing internal components
+    // It creates a new keccak matrix with placeholders instead of
+    // a default
+    fn new_keccakmatrix(&mut self) -> [Word64; 25] {
+        let mut matrix: [Word64; 25] = [Word64::default(); 25];
+        (0..25).for_each(|x| matrix[x] = self.new_word64());
+        matrix
+    }
+
+    // NOTE: this is only used for testing internal components
+    // It sets the wireId's in the keccak matrix to the provided
+    // values
+    fn set_keccakmatrix(&mut self, matrix_wires: &[Word64; 25], input: &[u64; 25]) {
+        matrix_wires
+            .iter()
+            .zip(input.iter())
+            .for_each(|(wrd64, num)| self.set_word64(&wrd64, *num));
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////
+    /////////////////////////// Const Word8 and Word64 /////////////////////////////
+    ////////////////////////////////////////////////////////////////////////////////
+
+    pub fn const_word8(&mut self, input: u8) -> Word8 {
+        let mut n = input;
```

Comment:
I keep forgetting that by taking ownership you "own" the type and can do whatever you want with it. Plus marking the input type as `mut` does not appear in the compiled docs because it does not mater. The only distinctions that matter are `&mut` and `&`. Change made (but without the `*` before the `mut` I assume that was a typo.